core/src/main/java/org/infinispan/interceptors/distribution/DistributionBulkInterceptor.java

```diff
@@ -226,17 +233,17 @@ public V setValue(V value) {
    @Override
    public CompletableFuture<Void> visitKeySetCommand(InvocationContext ctx, KeySetCommand command) throws Throwable {
       CacheSet<K> keySet;
-      if (!command.hasFlag(Flag.CACHE_MODE_LOCAL)) {
-         if (ctx.isInTxScope()) {
-            keySet = new TxBackingKeySet<>(Caches.getCacheWithFlags(cache, command),
-                  cache.getAdvancedCache().withFlags(Flag.CACHE_MODE_LOCAL).cacheEntrySet(), command,
-                  (LocalTxInvocationContext) ctx);
-         } else {
-            keySet = new BackingKeySet<>(Caches.getCacheWithFlags(cache, command), cache.getAdvancedCache().withFlags(
-                    Flag.CACHE_MODE_LOCAL).cacheEntrySet(), command);
-         }
+      if (command.hasFlag(Flag.CACHE_MODE_LOCAL)) {
+         return ctx.continueInvocation();
+      }
+
+      if (ctx.isInTxScope()) {
+         keySet = new TxBackingKeySet<>(Caches.getCacheWithFlags(cache, command),
+               cache.getAdvancedCache().withFlags(Flag.CACHE_MODE_LOCAL).cacheEntrySet(), command,
+               (LocalTxInvocationContext) ctx);
       } else {
-         keySet = (CacheSet<K>) ctx.forkInvocationSync(command);
+         keySet = new BackingKeySet<>(Caches.getCacheWithFlags(cache, command), cache.getAdvancedCache().withFlags(
+                 Flag.CACHE_MODE_LOCAL).cacheEntrySet(), command);
       }
       return ctx.shortCircuit(keySet);
```

Comment:
`visitKeySetCommand` uses a trick (`cache.cacheEntrySet()`) to invoke a completely different command synchronously without `forkInvocationSync()`, and needs to skip the rest of the interceptors for the `KeySetCommand`. We'd have to return a `CompletableFuture.completedFuture(keySet)` only if we executed the other interceptors and created the key set in a `ReturnHandler`.

Personally I'd rather remove the `KeySetCommand` and to this wrapping in `CacheImpl`, but this code only ever gets executed on the originator, and we don't have/need an async version of `Cache.keySet()`, so I decided to keep this interceptor as is.